polkadot/xcm/xcm-builder/src/barriers.rs

```diff
@@ -307,11 +310,32 @@ impl<T: Contains<Location>> ShouldExecute for AllowExplicitUnpaidExecutionFrom<T
 			origin, instructions, max_weight, _properties,
 		);
 		ensure!(T::contains(origin), ProcessMessageError::Unsupported);
-		instructions.matcher().match_next_inst(|inst| match inst {
-			UnpaidExecution { weight_limit: Limited(m), .. } if m.all_gte(max_weight) => Ok(()),
-			UnpaidExecution { weight_limit: Unlimited, .. } => Ok(()),
-			_ => Err(ProcessMessageError::Overweight(max_weight)),
-		})?;
+		// We will read up to 5 instructions. This allows up to 3 `ClearOrigin` instructions. We
+		// allow for more than one since anything beyond the first is a no-op and it's conceivable
+		// that composition of operations might result in more than one being appended.
+		let end = instructions.len().min(5);
+		instructions[..end]
+			.matcher()
+			// We allow one teleport or reserve asset transfer instruction before
+			// altering the origin and using `UnpaidExecution`.
+			// We reuse `MAX_ASSETS_FOR_BUY_EXECUTION` so we cap the number of assets that
+			// can be transferred this way in case `UnpaidExecution` fails.
+			.skip_next_inst_if(|inst| {
+				matches!(inst, ReceiveTeleportedAsset(assets) | ReserveAssetDeposited(assets) if assets.len() <= MAX_ASSETS_FOR_BUY_EXECUTION)
+			})?
+			// We then allow multiple instructions that alter the origin.
+			.skip_inst_while(|inst| {
+				matches!(inst, ClearOrigin | AliasOrigin(..)) ||
+					matches!(inst, DescendOrigin(child) if child != &Here) ||
+					matches!(inst, SetHints { .. })
+			})?
```

Comment:
We need to allow multiple asset transfer instructions, one of each should suffice. We also can't limit on `MAX_ASSETS_FOR_BUY_EXECUTION` because we need to be able to transfer in as many assets as the user wants before altering origin.
But that is ok: as long as this barrier explicitly checks if the altered `origin` is allowed to run `UnpaidExecution`, then we know that unpaid execution is actually allowed, the origin is trusted, and DoS is not a problem.

```suggestion
		instructions[..end]
			.matcher()
			// We skip all types of asset transfer instructions, skip origin altering instructions,
			// but expect explicit `UnpaidExecution` instruction.
			.skip_inst_while(|inst| {
				matches!(inst, ReceiveTeleportedAsset(_) | ReserveAssetDeposited(_) | WithdrawAsset(_) | SetHints { .. }) ||
					matches!(inst, ClearOrigin | AliasOrigin(..) | DescendOrigin(child) if child != &Here)
			})?
```

Reply:
I limited them since we don't check if `UnpaidExecution` is successful in the barrier, same as with `BuyExecution/PayFees`. To do that, we would need the barrier to process all origin altering instructions and then execute `UnpaidExecution` to know if it's good to go.

We could do that since it'll be cheaper than processing all assets in transfer instructions and then erroring.

Reply:
> To do that, we would need the barrier to process all origin altering instructions and then execute UnpaidExecution to know if it's good to go.

yes, exactly. This is what this barrier should do anyway. It is through the generic param of this barrier that the runtime configures which origins are allowed to do unpaid exec.

the executor doesn’t check anything extra. If it passes here it, “UnpaidExecution” will execute successfully.